RBD in-tree plugin migration to CSI driver using migration translation lib

humblec opened the pull request:
Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

**What type of PR is this?**

In support of csi-migration proposal here:
https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/csi-migration.md

this help with migration of in-tree RBD plugin ( kubernetes.io/rbd) to RBD CSI driver ( rbd.csi.ceph.com ).

Fixes #kubernetes/enhancements#2923

/kind feature
/kind sig-storage

Updates # https://github.com/kubernetes/kubernetes/issues/91246

-->
```release-note
 This release enables in-tree RBD migration to CSI driver with a couple of feature gates.  These featuregates are alpha in this release.

- `CSIMigrationRBD`: when enable, it will redirect traffic from in tree rbd plugin ( kubernetes.io/rbd )  to CSI driver ( rbd.csi.ceph.com) , default to `false` now.
- `IntreePluginRBDUnregister`: Disables the RBD in-tree driver

The featuregates can be enabled by:

1. Adding the feature flag to the kube-controller-manager `--feature-gates=CSIMigrationRBD=true` 
2. Adding the feature flag to the kubelet config:

featureGates:
  CSIMigrationRBD: true


As a Kubernetes cluster operator that administers storage, here are the prerequisites that you must complete before you attempt migration to the RBD CSI driver:

* You must install the Ceph CSI driver (rbd.csi.ceph.com), v3.5.0 or above, into your Kubernetes cluster.
* Considering the clusterID field is a required parameter for CSI driver for its operations, but in-tree StorageClass has monitors field as a required parameter, a Kubernetes storage admin has to create a clusterID based on the monitors hash ( ex:#echo -n '<monitors_string>' | md5sum) in the CSI config map and keep the monitors under this clusterID configuration.
* Also, if the value of adminId in the in-tree Storageclass is different from admin, the adminSecretName mentioned in the in-tree Storageclass has to be patched with the base64 value of the adminId parameter value, otherwise this step can be skipped.

 ```

fedebongio commented:
/remove-sig api-machinery

huchengze commented:
@humblec Hello! Are you still working on this? 

I found that to finish rbd csi migration, you not only need to update the code in kubernetes, but also some code in [external-provisioner](https://github.com/kubernetes-csi/external-provisioner) .

If you don't have enough time to do this, I can help.

humblec commented:
> @humblec Hello! Are you still working on this?
> 
> I found that to finish rbd csi migration, you not only need to update the code in kubernetes, but also some code in [external-provisioner](https://github.com/kubernetes-csi/external-provisioner) .
> 
> If you don't have enough time to do this, I can help.

@huchengze due to some personal emergency I was away so the delay.. Will be revisiting soon.

humblec commented:
/reopen

humblec commented:
/remove-lifecycle rotten

caesarxuchao commented:
/remove-sig api-machinery

humblec committed:
Add support for rbd plugin to csi-translation-lib

In support of csi-migration proposal here:
https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/csi-migration.md

Will help with migration of in-tree RBD plugin ( kubernetes.io/rbd)
to RBD CSI driver ( rbd.csi.ceph.com ).

Fixes https://github.com/kubernetes/enhancements/issues/2923

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

humblec commented:
/assign @Jiawei0227

humblec commented:
/kind feature
/kind api-change

humblec commented:
/test  pull-kubernetes-e2e-kind

humblec commented:
/assign @jsafrane 
/assign @msau42 
/assign @xing-yang

Jiawei0227 commented:
Thanks for your effort!
/lgtm
Overall looks good to me. I am not familiar enough with Ceph about the translation logic so I will defer it to @msau42 or @gnufied . But according to our offline discussion I think it pass all the e2e tests and you also had unit tests so should be good.

This PR does not involve API changes if I understand correctly? Also, can you add the new feature flags into the release notes and add a description for them. Thanks a lot!

/cc @msau42 for approval

fedebongio commented:
/remove-sig api-machinery

msau42 commented on staging/src/k8s.io/csi-translation-lib/plugins/rbd.go:
```diff
@@ -0,0 +1,313 @@
+}
+
+// fillDefaultSCParams fills some sc parameters with default values
+func fillDefaultSCParams(params map[string]string) {
```
Are these default values required for CSI but not required for intree?

humblec replied:
@msau42 yes it is required for csi and also intree fills a couple of  default values like pool which is handled here.

msau42 commented:
Also in the release notes, can you clarify that the feature is alpha?

humblec commented:
> Also in the release notes, can you clarify that the feature is alpha?

@msau42 done.

humblec commented:
/retest

humblec commented:
/test pull-kubernetes-unit

humblec commented:
/retest

humblec commented:
/test pull-kubernetes-e2e-gce-alpha-features

humblec commented:
/test pull-kubernetes-unit

humblec commented:
/retest

humblec commented:
/retest

humblec commented:
/retest

jsafrane commented on pkg/features/kube_features.go:
```diff
@@ -330,6 +330,18 @@ const (
+	// alpha: v1.23
+	//
+	// Enables the RBD in-tree driver to RBD CSI Driver  migration feature.
+	CSIMigrationRbd featuregate.Feature = "CSIMigrationRbd"
```
`CSIMigrationRBD`
(abbreviations have all the same case, like CSI)

humblec replied:
sure, done

jsafrane commented:
/retest
e2e-gce-alpha-features failure looks unrelated, just in case...

humblec commented on staging/src/k8s.io/csi-translation-lib/plugins/rbd.go:
```diff
@@ -0,0 +1,326 @@
+)
+
+const (
+	RbdVolumePluginName           = "kubernetes.io/rbd"
+	RbdDriverName                 = "rbd.csi.ceph.com"
```
sure, let me do that. thanks.

humblec commented on pkg/features/kube_features.go:
```diff
@@ -330,6 +330,18 @@ const (
+	// alpha: v1.23
+	//
+	// Disables the RBD in-tree driver.
+	InTreePluginRbdUnregister featuregate.Feature = "InTreePluginRbdUnregister"
```
this is addressed. Thanks .

humblec commented on staging/src/k8s.io/csi-translation-lib/plugins/rbd.go:
```diff
@@ -0,0 +1,326 @@
+	defaultIntreeImagePfx         = "kubernetes-dynamic-pvc-"
+	defaultMigKey                 = "migration"
+	defaultMigStaticVal           = "true"
+	CSIVolHandleAnnKey            = "rbd.csi.ceph.com/volume-handle"
```
done.

humblec commented:
> /retest e2e-gce-alpha-features failure looks unrelated, just in case...

yep, it looks like its failing for all the PRs.. Thanks @jsafrane

humblec commented:
/retest

jsafrane commented:
/lgtm
/approve
(the storage parts)

Jiawei0227 commented:
/assign @msau42 
for approval. Can you also help to tag the right folks for sig-node and sig-scheduler as well as kube-controller-manager change?

msau42 commented:
/approve
/assign @damemi @lavalamp @random-liu

humblec commented:
@msau42 @Jiawei0227 @xing-yang do we have to add any extra labels to set the priority or urgency against the code freeze ?

humblec commented:
/priority-backlog

humblec commented:
/priority backlog

ahg-g commented:
/approve
/hold 

@msau42 please unhold once the feature gate is added to https://github.com/kubernetes/kubernetes/blob/ed3d9f7b48290d26ea617b2af40188e397987ac6/pkg/scheduler/framework/plugins/volumebinding/binder.go#L1018

humblec commented:
> /approve /hold
> 
> @msau42 please unhold once the feature gate is added to
> 
> https://github.com/kubernetes/kubernetes/blob/ed3d9f7b48290d26ea617b2af40188e397987ac6/pkg/scheduler/framework/plugins/volumebinding/binder.go#L1018

@ahg-g @msau42 addressed above, ptal.. thanks in advance. :+1:

sttts commented:
/approve

for the kcm part

msau42 commented:
/lgtm

humblec commented:
/retest pull-kubernetes-e2e-gce-storage-slow

humblec commented:
/test  pull-kubernetes-e2e-gce-storage-slow

msau42 commented:
/hold cancel

humblec commented:
Demo : part 1 shows PVCs created before enabling migration and part2 shows PVCs created after migration.
https://drive.google.com/file/d/1yfcngCEyrOPDG5tOg7gva0wVJqNELRGc/view
